File: backend/app/models/questionnaire.py
from sqlalchemy import Column, Integer, String, Float, Boolean, DateTime, ForeignKey, JSON, Enum as SQLEnum
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
import enum
from app.db.session import Base


class QuestionnaireStatus(str, enum.Enum):
    """Questionnaire status"""
    DRAFT = "draft"
    SUBMITTED = "submitted"
    REVIEWED = "reviewed"


class Questionnaire(Base):
    """Patient questionnaire responses"""
    __tablename__ = "questionnaires"

    id = Column(Integer, primary_key=True, index=True)
    patient_id = Column(Integer, ForeignKey("users.id"), nullable=True)  # Nullable for anonymous submissions
    status = Column(SQLEnum(QuestionnaireStatus), default=QuestionnaireStatus.DRAFT)

    # Section I: Basic Information
    age = Column(Integer, nullable=False)
    gender = Column(String, nullable=False)
    contact_number = Column(String, nullable=True)
    is_childbearing_age_woman = Column(Boolean, nullable=True)
    height_ft = Column(Integer, nullable=False)
    height_in = Column(Integer, nullable=False)
    weight_lb = Column(Float, nullable=False)
    bmi = Column(Float, nullable=True)  # Calculated

    # Section II: Eating Habits & Feelings
    eating_habits = Column(JSON, nullable=True)  # List of eating habits

    # Section III: Medical Conditions & Health Status
    health_conditions = Column(JSON, nullable=True)  # List of health conditions

    # Section IV: Medication and Allergy History
    current_medications = Column(JSON, nullable=True)  # List of medications
    has_drug_allergies = Column(Boolean, nullable=False)
    drug_allergies = Column(JSON, nullable=True)  # List of allergies

    # Section V: Additional Remarks
    additional_remarks = Column(String, nullable=True)

    # Metadata
    submitted_at = Column(DateTime(timezone=True), nullable=True)
    reviewed_at = Column(DateTime(timezone=True), nullable=True)
    reviewed_by_doctor_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    # Relationships commented out to avoid SQLAlchemy ambiguous foreign key errors
    # Access related data using foreign key columns directly (patient_id, reviewed_by_doctor_id)
    # patient = relationship("User", foreign_keys=[patient_id])
    # reviewed_by = relationship("User", foreign_keys=[reviewed_by_doctor_id])
    # screening_result = relationship("ScreeningResult", foreign_keys="ScreeningResult.questionnaire_id", uselist=False)

File: ios/Models/Questionnaire.swift
import Foundation

struct QuestionnaireCreate: Codable {
    let age: Int
    let gender: String
    let contactNumber: String?
    let isChildbearingAgeWoman: Bool
    let heightFt: Int
    let heightIn: Int
    let weightLb: Double
    let eatingHabits: [String]
    let healthConditions: [String]
    let currentMedications: [String]
    let hasDrugAllergies: Bool
    let drugAllergies: [String]
    let additionalRemarks: String?
    
    enum CodingKeys: String, CodingKey {
        case age, gender
        case contactNumber = "contact_number"
        case isChildbearingAgeWoman = "is_childbearing_age_woman"
        case heightFt = "height_ft"
        case heightIn = "height_in"
        case weightLb = "weight_lb"
        case eatingHabits = "eating_habits"
        case healthConditions = "health_conditions"
        case currentMedications = "current_medications"
        case hasDrugAllergies = "has_drug_allergies"
        case drugAllergies = "drug_allergies"
        case additionalRemarks = "additional_remarks"
    }
}

struct Questionnaire: Codable {
    let id: Int?
}

struct ScreeningResult: Codable {
    let id: Int
    let questionnaireId: Int
    let isEligible: Bool
    let eligibilityMessage: String?
    let age: Int?
    let gender: String?
    let bmiCategory: String?
    let recommendedDrugs: [MedicationRecommendation]
    let warnings: [String]
    
    enum CodingKeys: String, CodingKey {
        case id
        case questionnaireId = "questionnaire_id"
        case isEligible = "is_eligible"
        case eligibilityMessage = "eligibility_message"
        case age, gender
        case bmiCategory = "bmi_category"
        case recommendedDrugs = "recommended_drugs"
        case warnings
    }
}

struct MedicationRecommendation: Codable {
    let medication: String
    let priority: Int
    let reasoning: String
}
